src/Uno.UI/UI/Xaml/XamlRoot.skia.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Diagnostics;
+using Uno.Foundation.Logging;
+using Uno.UI.Dispatching;
+
+namespace Windows.UI.Xaml;
+
+public sealed partial class XamlRoot
+{
+	private bool _isMeasureWaiting = false;
+	private bool _isArrangeWaiting = false;
+	private bool _isMeasuringOrArranging = false;
+	private bool _renderQueued = false;
+	
+	internal event Action InvalidateRender = () => { };
+
+	internal void InvalidateMeasure() => ScheduleInvalidateMeasureOrArrange(invalidateMeasure: true);
+
+	internal void InvalidateArrange() => ScheduleInvalidateMeasureOrArrange(invalidateMeasure: false);
+
+	internal void QueueInvalidateRender()
+	{
+		if (!_isMeasuringOrArranging && !_renderQueued)
+		{
+			_renderQueued = true;
+
+			CoreDispatcher.Main.RunAsync(CoreDispatcherPriority.Normal, () =>
+			{
+				_renderQueued = false;
+				InvalidateRender();
+			});
+		}
+	}
+
+	internal void ScheduleInvalidateMeasureOrArrange(bool invalidateMeasure)
+	{
+		if (invalidateMeasure)
+		{
+			if (_isMeasureWaiting)
+			{
+				// A measure is already queued
+				return;
+			}
+
+			_isMeasureWaiting = true;
+
+			if (_isArrangeWaiting)
+			{
+				// Since an arrange is already queued, no need to
+				// schedule something on the dispatcher
+				return;
+			}
+		}
+		else
+		{
+			if (_isArrangeWaiting)
+			{
+				// An arrange is already queued
+				return;
+			}
+
+			_isArrangeWaiting = true;
+
+			if (_isMeasureWaiting)
+			{
+				// Since a measure is already queued, no need to
+				// schedule something on the dispatcher
+				return;
+			}
+		}
+
+		CoreDispatcher.Main.RunAsync(CoreDispatcherPriority.Normal, () => RunMeasureAndArrange());
+	}
+
+	private void RunMeasureAndArrange()
```

Comment:
Is this coming from existing code / WInUI code? I would expect a loop of 250 tries before the infamous 'LayoutCycleException`

Reply:
This code was moved from `Window.skia.cs`, with little to no changes, hopefully should not be breaking anything 😀